### PR TITLE
feat(api): Implement extension API and nvim_tree extension

### PR DESCRIPTION
**Pull request by mrjones2014**
Resolves: #183 

## How to Test

1. Load the plugin from this branch, and setup with `require('legendary').setup({ extensions = { nvim_tree = true } })`
2. Make sure nvim-tree is installed
3. Open nvim-tree
4. Trigger the legendary.nvim finder from the nvim-tree window
5. It should show the nvim-tree keymaps
6. Triggering the legendary.nvim finder from _any other window_ should _not_ show the nvim-tree keymaps

## Testing for Regressions

I have tested the following:

- [x] Triggering keymaps from `legendary.nvim` in all modes (normal, insert, visual)
- [x] Creating keymaps via `legendary.nvim`, then triggering via the keymap in all modes (normal, insert, visual)
- [x] Triggering commands from `legendary.nvim` in all modes (normal, insert, visual)
- [x] Creating commands via `legendary.nvim`, then running the command manually from the command line
- [x] `augroup`/`autocmd`s created through `legendary.nvim` work correctly

### CODE REVIEW

**Comment by mrjones2014** [comment]
@alex-courtis Could you give this a quick review and make sure I'm not doing anything stupid or unsupported with the nvim-tree API?

@hinell could you also provide some thoughts and feedback?

**Comment by mrjones2014** [comment]
@alex-courtis The idea here is that `legendary.nvim` users can now just specify this field in their config:

```lua
require('legendary').setup({
  extensions = {
    nvim_tree = true,
  },
})
```

And the `nvim-tree` keymaps will show up when appropriate in the `legendary.nvim` finder.

**Comment by alex-courtis** [comment]
This is really nice; almost nothing required from the user.

Any plans to apply keymaps to nvim-tree?

File: lua/legendary/extensions/nvim_tree.lua [hunk 1]
```diff
@@ -0,0 +1,25 @@
+local function init()
```
Comment (alex-courtis) [inline lua/legendary/extensions/nvim_tree.lua:1]: We're waiting until the user has called setup, so user mappings will be present.

**Comment by mrjones2014** [comment]
> Any plans to apply keymaps to nvim-tree on_attach?

what do you mean, exactly? Isn’t this handled by nvim-tree?

**Comment by mrjones2014** [comment]
> > Any plans to apply keymaps to nvim-tree on_attach?
> 
> what do you mean, exactly? Isn’t this handled by nvim-tree?

@alex-courtis

File: lua/legendary/extensions/nvim_tree.lua [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+  end, keymaps)
+  require('legendary').keymaps(legendary_keymaps)
+  require('legendary').commands({
+    { 'NvimTreeOpen', description = 'nvim-tree: Open file tree' },
+    { 'NvimTreeClose', description = 'nvim-tree: Close file tree' },
+    { 'NvimTreeToggle', description = 'nvim-tree: Toggle file tree' },
+    { 'NvimTreeFocus', description = 'nvim-tree: Focus window' },
+    { 'NvimTreeRefresh', description = 'nvim-tree: Refresh file tree' },
+  })
```
Comment (mrjones2014) [inline lua/legendary/extensions/nvim_tree.lua:18]: also @alex-courtis is there a way to get these without hard-coding them?

Comment (mrjones2014) [reply lua/legendary/extensions/nvim_tree.lua:19]: opened https://github.com/nvim-tree/nvim-tree.lua/issues/2076

**Comment by alex-courtis** [comment]
> > Any plans to apply keymaps to nvim-tree on_attach?
> 
> what do you mean, exactly? Isn’t this handled by nvim-tree?

Yes, just curious about your future plans...

**Comment by hinell** [comment]
Just remember: autocmds are preferred. That's all.

**Comment by mrjones2014** [comment]
> autocmds are preferred

could you elaborate

**Comment by hinell** [comment]
@mrjones2014 I mean, if you want to bind keymaps to a buffer you should use autocmds instead of the nvim-trees' `on_attach` config.

**Pull request merged**